### ISSUE TITLE
Higher-level API

### ISSUE THREAD

**Issue by mikea** [google/libprotobuf-mutator#4]
We discussed a bit how high-level api to the library could look like. My opinion is that it should look like (please ignore names):

```c++

extern "C" int LLVMFuzzerTestProto(const MyProto& proto) {
  // put your fuzzing code here and use proto as input.
  return 0;
}

// Or even
LLVM_FUZZER_PROTO(MyProto proto) {
  // put your fuzzing code here and use proto as input.
  return 0;
}

// Or maybe with registrar:
extern "C" int MyFuzzerTestProto(const MyProto& proto) {
  // put your fuzzing code here and use proto as input.
  return 0;
}
REGISTER_PROTO_FUZZER(MyFuzzerTestProto);


```

The important point is: the same way libfuzzer api doesn't introduce too many knowledge about fuzzing engine, this shouldn't either. If I don't care about custom proto mutations - I shouldn't bother writing code for it.

**Comment by kcc** [comment]
I agree we should do something like this, but not sure how.
Your 3 variants all look sane, the first one is better, if we can implement it. 
Before coming up with a 'final' interface I'd like to have 2-3 examples in this repo using the existing (verbose) interface. 

One question: what headers are we going to require? 
Obviously, we will need `#include "myproto.pb.h"`, but what else? 
I don't want to require any headers from libFuzzer. 
And if possible not even from  libprotobuf-mutator

> `use data+size as input`
did you mean `use proto as input`?

**Comment by mikea** [comment]
I don't know if not including mutator header is feasible. We discussed this at a length and we didn't find a way to discover function with unknown signature. 

But this brings a related question to my mind: maybe we want mutator library to be header-only?

PS fixed code samples comments

**Comment by kcc** [comment]
> maybe we want mutator library to be header-only?

That's possible and might be a part of solution.  
But then again, what if we want to use some other proto mutation tool? 

> I don't know if not including mutator header is feasible. 

Well, at the very least we can use the -i compiler flag to include something under the hood for cases were we control the build system. 

Anyway, let's not rush here. With more proto fuzz targets a better understanding may come.

**Comment by vitalybuka** [comment]
We should not use header only, as it will force us to compile and instrument mutator together with fuzz target.

If we want to avoid headers at all, we can go with:
```
extern "C" google::protobuf::Message* LLVMFuzzerGetDefaultProto() {
  return MyProto::default_instance();
}
extern "C" int LLVMFuzzerTestProto(google::protobuf::Message& message) {
  const MyProto& my_message = static_cast<const MyProto&>(message);
  ....
  return 0;
}
```
So user will have to cast.

If we allow some small header, I'd go with
```
void SomeTestCode(const MyProto& message) {
}
protobuf_mutator::libfuzzer::TextFuzzer<MyProto> fuzzer(&SomeTestCode); // or just lambda
```

Which can be reduced to macro, like in the first comment: 
```
LLVM_FUZZER_PROTO(const MyProto& proto) {
  // Emulate a bug.
}
```

**Comment by vitalybuka** [comment]
Fixed with libprotobuf-mutator/pull/63

**Comment by vitalybuka** [comment]
libprotobuf-mutator/pull/63
